TIFF.pas

```diff
@@ -621,26 +621,49 @@ procedure GraphicExTIFFError(Module: PChar; const Format: PChar; Params: va_list
 
 var
   Buffer: array[0..1000] of Char;
-
+//  Param1, Param2: PAnsiChar;
+//  wtf: AnsiString;
 begin
+//try to find out what's wrong with Params
+//  Param1:=Params^;
+//  Param2:=Param1+1;
+//  wtf:=Param1;
+//  GraphicExError(SysUtils.Format(Format,[Param1, Param2]));
+//  Param2:=PAnsiChar((Params+1)^);
   wvsprintf(Buffer, Format, Params);
   GraphicExError(Buffer);
 end;
 
 //----------------------------------------------------------------------------------------------------------------------
 
 procedure GraphicExTIFFWarning(Module: PChar; const Format: PChar; Params: va_list); cdecl;
-
 var
   Buffer: array[0..1000] of Char;
-
 begin
   wvsprintf(Buffer, Format, Params);
   OutputDebugString(Buffer);
 end;
 
 //----------------------------------------------------------------------------------------------------------------------
 
+function zcalloc(opaque: Pointer; items, size: Integer): Pointer;
+begin
+  GetMem(result,items * size);
+end;
+
+procedure zcfree(opaque, block: Pointer);
+begin
+  FreeMem(block);
+end;
+
+
+{$ifndef WIN64}
+procedure _llmod;
+asm
+  jmp System.@_llmod;
```

Comment:
Did you check that the call stack is properly cleaned up and the calling convention is correct? This is code called from a C library, so it might expect a different calling convention. It's very important that the stack is cleaned properly or you easily get a security problem.